### PR TITLE
[WIP] Polygon and Circle Detection

### PR DESCRIPTION
**Pull request by neel-j-dev**
#### FindPolygonsPipe

- Input: `List<Contour>`
- Ouput:` List<CVShape>`

#### FilterPolygonsPipe
- Takes in List<CVShape> and outputs filtered List<CVShape> based off perimeter, area, and desired shape

#### FindCirclesPipe
- Input: `Pair<Mat, List<Contour>>`
- Ouput: `List<CVShape>`

### Pipelines
- Fleshed out ColouredShapePipeline

### Testing
Not too much testing done yet. TODOs for testing include:
- [x]  Use powercell images from WPI's ML dataset
- [x] Use more FRC specific targets instead of perfect polygons

### Triangles
image

### Quadrilaterals 
image

### Circles
image

General Things left

- [x] Test more
- [x] Replace draw2dcontours pipe with draw2dshapes pipe
- [x] Add solvepnp support
- [x] Make HoughCircles params a part of FindCirclesPipe params

### Results from power cell test images
image
image
image

Higher resolution and less blur give better results and you might need to play around with the cannyMaxThresh and accuracy to pick up circles and not duck tape/similar shapes wires. But filterShapesPipe should take care of it if you want to define a min Area, peri or the minDist, min/max radius in the FindCirclespipe.

### Solvepnp

Tried to draw a cube enclosing the circle, but ran into a wall with that since drawing over a contour made it impossible. Instead, a single square worked nicely. 
image

### CODE REVIEW

File: chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    private final HSVPipe hsvPipe = new HSVPipe();
+    private final OutputMatPipe outputMatPipe = new OutputMatPipe();
+    private final FindContoursPipe findContoursPipe = new FindContoursPipe();
+    private final SpeckleRejectPipe speckleRejectPipe = new SpeckleRejectPipe();
+    private final GroupContoursPipe groupContoursPipe = new GroupContoursPipe();
+    private final SortContoursPipe sortContoursPipe = new SortContoursPipe();
+    private final Collect2dTargetsPipe collect2dTargetsPipe = new Collect2dTargetsPipe();
+    private final Draw2dCrosshairPipe draw2dCrosshairPipe = new Draw2dCrosshairPipe();
+    private final Draw2dContoursPipe draw2dContoursPipe = new Draw2dContoursPipe();
+
```
Comment (Bankst) [inline chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java]: Just a nitpick. List the pipes in the order they used

File: chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    private Mat rawInputMat = new Mat();
+    private DualMat outputMats = new DualMat();
+
+    ColouredShapePipeline() {
+        settings = new ColouredShapePipelineSettings();
+    }
+
+    @Override
+    protected void setPipeParams(
```
Comment (Bankst) [inline chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java]: Same nitpick here. Set up the params in the pipe order.

File: chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        sumPipeNanosElapsed += outputMatResult.nanosElapsed;
+
+        CVPipeResult<List<Contour>> findContoursResult = findContoursPipe.apply(hsvPipeResult.result);
+        sumPipeNanosElapsed += findContoursResult.nanosElapsed;
+
+        CVPipeResult<List<Contour>> speckleRejectResult =
+                speckleRejectPipe.apply(findContoursResult.result);
+        sumPipeNanosElapsed += speckleRejectResult.nanosElapsed;
+
+        List<CVShape> shapes;
```
Comment (Bankst) [inline chameleon-server/src/main/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipeline.java]: Move this out of the loop and reuse it each time

File: chameleon-server/src/test/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipelineTest.java [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+import com.chameleonvision.common.vision.opencv.ContourIntersectionDirection;
+import com.chameleonvision.common.vision.opencv.ContourShape;
+import java.io.File;
+import java.util.Objects;
+import org.junit.jupiter.api.Test;
+
+public class ColouredShapePipelineTest {
+
+    @Test
+    public static void testTriangleDetection(
```
Comment (Bankst) [inline chameleon-server/src/test/java/com/chameleonvision/common/vision/pipeline/ColouredShapePipelineTest.java]: These methods don't need the `@Test` attribute if they don't assert anything.

**Review by Bankst — CHANGES_REQUESTED** [review]
Get your `.gradle` folder in to `.gitignore`.
Make the file naming of your test images a common scheme

**Comment by mcm001** [comment]
Can you uncommit the stuff in `.gradle`?

**Review by Bankst — APPROVED** [review]

**Pull request merged**